### DISCUSSION TITLE
psycopg2.OperationalError due to idle_in_transaction_session_timeout goes unhandled

### DISCUSSION THREAD

**Discussion by dilshans2k** [sqlalchemy/sqlalchemy#11617] [Usage Questions]
### Describe the bug

I require that a transaction started must not exceed certain time in idle (not commited or rollback case). **SQLAlchemy** provides `idle_in_transaction_session_timeout` which will remove the transaction from `idle_in_transaction` state in PgSQL which works as expected, but on the application side (**Flask**) i face two issues:
1. The error thrown due to timeout does not happen at the instant when timeout was triggered, but only after the blocking work was done.
2. The error thrown is not being caught by any exception handler, neither try-except block, nor flask's global exception handler (registered at time of app init).

### SQLAlchemy Version in Use

2.0.31

### DBAPI (i.e. the database driver)

psycopg2

### Database Vendor and Major Version

PostgreSQL 16

### Python Version

3.12

### Operating system

Linux

### To Reproduce

```python
from sqlalchemy import create_engine, text
from sqlalchemy.orm import sessionmaker, scoped_session, DeclarativeBase
import time
import logging
from sqlalchemy import Column, Integer

class Base(DeclarativeBase):
    pass

class Citation(Base):

    """citation count model."""

    __tablename__ = "citation_count"

    id = Column(Integer, primary_key=True)

logging.getLogger().setLevel(logging.DEBUG)

idle_in_transaction_session_timeout = 5000

engine = create_engine(
    f"postgresql://pguser:pguser@localhost:5454/pgdb1?options=-c%20idle_in_transaction_session_timeout={idle_in_transaction_session_timeout}",
    echo=True,
)
Session = sessionmaker(bind=engine)
session = scoped_session(Session)

try:
    session.query(Citation).filter().all()
    # Some long running task
    time.sleep(6)
except Exception as e:
    # Never printed
    print("Exception received: ", e)
```

### Error

```
2024-07-18 13:48:47,559 INFO sqlalchemy.engine.Engine select pg_catalog.version()
2024-07-18 13:48:47,560 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-07-18 13:48:47,564 INFO sqlalchemy.engine.Engine select current_schema()
2024-07-18 13:48:47,564 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-07-18 13:48:47,566 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2024-07-18 13:48:47,567 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-07-18 13:48:47,569 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-07-18 13:48:47,574 INFO sqlalchemy.engine.Engine SELECT citation_count.id AS citation_count_id 
FROM citation_count
2024-07-18 13:48:47,574 INFO sqlalchemy.engine.Engine [generated in 0.00034s] {}
Exception during reset or similar
Traceback (most recent call last):
  File "/home/dilshan.sandhu/Projects/grandslam/dxassist/.venv/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 986, in _finalize_fairy
    fairy._reset(
  File "/home/dilshan.sandhu/Projects/grandslam/dxassist/.venv/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 1432, in _reset
    pool._dialect.do_rollback(self)
  File "/home/dilshan.sandhu/Projects/grandslam/dxassist/.venv/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 698, in do_rollback
    dbapi_connection.rollback()
psycopg2.OperationalError: server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
``` 

### Additional context

I am aware that using `autocommit=True` will ensure timeout is never triggered, but my use case does not allow it.
I have added a basic reproduction in a scripting manner. Let me know if Flask implementation is required, but stack traces are similar nonethless.

**Comment by zzzeek** [comment]
hi -

When the PostgreSQL server closes out the connection, there is nothing actively transmitted nor is there any kind of thread running in SQLAlchemy or in the database driver that could perform an "interrupt" of some kind here.     If your python application is sleeping, that's all it's going to do, there's no way in Python to have that suddenly interrupted based on some data waiting on a TCP socket, unless you organized for polling of some kind within your sleep.

**Reply by dilshans2k** [reply]
> hi -
> 
> When the PostgreSQL server closes out the connection, there is nothing actively transmitted nor is there any kind of thread running in SQLAlchemy or in the database driver that could perform an "interrupt" of some kind here. If your python application is sleeping, that's all it's going to do, there's no way in Python to have that suddenly interrupted based on some data waiting on a TCP socket, unless you organized for polling of some kind within your sleep.

Hi,
That resolves my first issue regarding why the application cannot notify about a database connection error while performing an action. However, my second issue remains unresolved: why isn't the error captured by the exception handler (try-except) or even by Flask's exception handler?
Is there any way to handle this scenario?
Thanks

**Reply by zzzeek** [reply]
this is the same thing, no exception is thrown in your code.  the message in your standard out is an exception that was caught, handled, and logged by the connection pool as the program is terminating, and the pool is attempting to clean up connections that were present in the pool.  this all happens after your exception block.

> Is there any way to handle this scenario?

the actual scenario isn't clear to me since this is a Flask application, it means you're serving HTTP requests.    IMO a correctly architected program would never be a need to use a setting like "idle transaction timeout", in an HTTP application, database transactions should last only up to a few hundred milliseconds at the most, and you'd be following patterns described at https://docs.sqlalchemy.org/en/20/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it : "basic pattern is create a [Session](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session) at the start of a web request, call the [Session.commit()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.commit) method at the end of web requests that do POST, PUT, or DELETE, and then close the session at the end of web request."
